sw/device/lib/dif/dif_alert_handler.c

```diff
@@ -4,9 +4,930 @@
+        break;
+      case kDifAlertHandlerClassStatePhase1:
+        enable_bit = ALERT_HANDLER_CLASSA_CTRL_EN_E1;
+        map_field.mask = ALERT_HANDLER_CLASSA_CTRL_MAP_E1_MASK;
+        map_field.index = ALERT_HANDLER_CLASSA_CTRL_MAP_E1_OFFSET;
+        break;
+      case kDifAlertHandlerClassStatePhase2:
+        enable_bit = ALERT_HANDLER_CLASSA_CTRL_EN_E2;
+        map_field.mask = ALERT_HANDLER_CLASSA_CTRL_MAP_E2_MASK;
+        map_field.index = ALERT_HANDLER_CLASSA_CTRL_MAP_E2_OFFSET;
+        break;
+      case kDifAlertHandlerClassStatePhase3:
+        enable_bit = ALERT_HANDLER_CLASSA_CTRL_EN_E3;
+        map_field.mask = ALERT_HANDLER_CLASSA_CTRL_MAP_E3_MASK;
+        map_field.index = ALERT_HANDLER_CLASSA_CTRL_MAP_E3_OFFSET;
+        break;
+      default:
+        return false;
+    }
+
+    ctrl_reg = bitfield_bit32_write(ctrl_reg, enable_bit, true);
+    ctrl_reg = bitfield_field32_write(ctrl_reg, map_field,
+                                      class->phase_signals[i].signal);
+  }
+
+  mmio_region_write32(handler->params.base_addr, reg_offset, ctrl_reg);
+
+  switch (class->alert_class) {
+    case kDifAlertHandlerClassA:
+      reg_offset = ALERT_HANDLER_CLASSA_ACCUM_THRESH_REG_OFFSET;
+      break;
+    case kDifAlertHandlerClassB:
+      reg_offset = ALERT_HANDLER_CLASSB_ACCUM_THRESH_REG_OFFSET;
+      break;
+    case kDifAlertHandlerClassC:
+      reg_offset = ALERT_HANDLER_CLASSC_ACCUM_THRESH_REG_OFFSET;
+      break;
+    case kDifAlertHandlerClassD:
+      reg_offset = ALERT_HANDLER_CLASSD_ACCUM_THRESH_REG_OFFSET;
+      break;
+    default:
+      return false;
+  }
+  mmio_region_write32(handler->params.base_addr, reg_offset,
+                      class->accumulator_threshold);
+
+  switch (class->alert_class) {
+    case kDifAlertHandlerClassA:
+      reg_offset = ALERT_HANDLER_CLASSA_TIMEOUT_CYC_REG_OFFSET;
+      break;
+    case kDifAlertHandlerClassB:
+      reg_offset = ALERT_HANDLER_CLASSB_TIMEOUT_CYC_REG_OFFSET;
+      break;
+    case kDifAlertHandlerClassC:
+      reg_offset = ALERT_HANDLER_CLASSC_TIMEOUT_CYC_REG_OFFSET;
+      break;
+    case kDifAlertHandlerClassD:
+      reg_offset = ALERT_HANDLER_CLASSD_TIMEOUT_CYC_REG_OFFSET;
+      break;
+    default:
+      return false;
+  }
+  mmio_region_write32(handler->params.base_addr, reg_offset,
+                      class->irq_deadline_cycles);
+
+  return true;
+}
+
+/**
+ * Configures phase durations of a particular class.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool configure_phase_durations(
+    const dif_alert_handler_t *handler,
+    const dif_alert_handler_class_config_t *class) {
+  if (class->phase_durations == NULL && class->phase_durations_len != 0) {
+    return false;
+  }
+
+  for (int i = 0; i < class->phase_durations_len; ++i) {
+    // To save on writing a fairly ridiculous `if` chain, we use a lookup table
+    // that leverages the numeric values of enum constants.
+    static const ptrdiff_t kRegOffsets[4][4] = {
```

Comment:
Magic number `4`, please document or create a named define/constant.

Reply:
I don't believe the second one has a convenient constant anywhere, so I'll just document it.